tests/plugins/test_ddp_spawn_plugin.py

```diff
@@ -77,3 +78,22 @@ def test_ddp_spawn_extra_parameters(tmpdir):
     trainer.fit(model, datamodule=dm)
     assert trainer.callback_metrics[val_name] == torch.tensor(val)
     assert model.test_val == "test_val"
+
+
+class BoringModelDDP(BoringModel):
+    def on_train_start(self) -> None:
+        """Check model wrapped by DistributedDataParallel at configurate_ddp."""
+        assert isinstance(self.trainer.model, DistributedDataParallel)
+
+    def on_fit_end(self) -> None:
+        """Check teardown function at the end of training."""
+        assert self.trainer.model.device == torch.device("cpu")
+        assert isinstance(self.trainer.model, LightningModule)
```

Comment:
don't we want these asserts in the main test below after fit?

Reply:
Yeah, that will work too! Could you share some insight about which way is better? I think they are the same logic, is that having test in main test the preferred way? thanks~~

Reply:
I think it's better to directly do the assert after the fit call because

1) nothing guarantees us here that on_fit_end actually runs. It's supposed to be a hook yes, but as far as the test is concerned here we have no way of knowing. it the hook doesn't run, the test will still pass!
2) I would not rely on the teardown/on_fit_end order to be the same in the future. especially since the recent changes where sometimes the teardown runs after on_fit_end and sometimes before. cc @ananthsub 
3) also if we do it directly in the main test we don't need this special model class just for testing

Reply:
> I would not rely on the teardown/on_fit_end order to be the same in the future. especially since the recent changes where sometimes the teardown runs after on_fit_end and sometimes before

IIRC `teardown` should always happen after `on_fit_end` : https://github.com/PyTorchLightning/pytorch-lightning/blob/538e743f17c7da4624c902f762922e2837661818/pytorch_lightning/plugins/training_type/ddp_spawn.py#L208-L214

@awaelchli do you see this isn't the case somewhere? how should we enforce this as an invariant of the loop?

Reply:
Yes @ananthsub for ddp spawn _call_teardown_hook runs before on_fit_end [(trainer.py line 1011)](https://github.com/PyTorchLightning/pytorch-lightning/blob/de57feff445744f431eaef29182b1bcbf58a1924/pytorch_lightning/trainer/trainer.py#L1017) and for anything not ddp_spawn it runs after. 

I can't find the PR anymore but these are repeated patterns that we have seen in the past and that will happen in the future, because PL community and PL team cannot agree if they want consistent hook call order or plugins that can do whatever they want.

Reply:
I see. this is the same question about why `on_fit_start` is not part of the loop. IMO consistency here is more critical. Let's file a separate issue/discussion to track this